### PR TITLE
⚠️ Update references with an APIReader as fallback

### PR DESCRIPTION
**Pull request by sbueringer**
Signed-off-by: Stefan Büringer buringerst@vmware.com

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Implements first part of https://github.com/kubernetes-sigs/cluster-api/issues/5686

### CODE REVIEW

File: util/conversion/conversion.go [hunk 1]
```diff
@@ -90,6 +93,44 @@ func UpdateReferenceAPIContract(ctx context.Context, c client.Client, ref *corev
 	return nil
 }
 
+// UpgradeReferenceAPIContract takes a client and object reference, queries the API Server for
+// the Custom Resource Definition and looks which one is the stored version available.
+//
+// The object passed as input is modified in place if an updated compatible version is found.
+// NOTE: In case CRDs are named incorrectly, this func is using an APIReader to list CRDs
+// to avoid to implicitly create an informer for CRDs which would leads to high memory consumption.
```
Comment (sbueringer) [inline util/conversion/conversion.go:101]: NOTE: We will have to do this deprecation cycle once more when we drop the apiReader.

Comment (fabriziopandini) [reply util/conversion/conversion.go:101]: I think we can do breaking changes in go code between minors, so we can drop previous signature

Comment (sbueringer) [reply util/conversion/conversion.go:101]: > I think we can do breaking changes in go code between minors, so we can drop previous signature

Do you mean changing it directly now or dropping the old one with v1.2.x?

**Note**: The signature we eventually need once we drop the fallback is this one: `func UpdateReferenceAPIContract(ctx context.Context, c client.Client, ref *corev1.ObjectReference) error`. So if we have a way to do a breaking change to this func directly without cycling through other func names I would prefer not deprecating it at all.

Comment (fabriziopandini) [reply util/conversion/conversion.go:101]: @vincepri  ^^

File: util/util.go [hunk 2]
```diff
@@ -432,6 +432,9 @@ func GetGVKMetadata(ctx context.Context, c client.Client, gvk schema.GroupVersio
 // GetCRDWithContract retrieves a list of CustomResourceDefinitions from using controller-runtime Client,
 // filtering with the `contract` label passed in.
 // Returns the first CRD in the list that matches the GroupVersionKind, otherwise returns an error.
+// Deprecated: This func is going to be removed in a next release.
+// NOTE: This func is using a client.Client to list CRDs and thus implicitly creates an informer for CRDs
+// which leads to high memory consumption.
 func GetCRDWithContract(ctx context.Context, c client.Client, gvk schema.GroupVersionKind, contract string) (*apiextensionsv1.CustomResourceDefinition, error) {
```
Comment (vincepri) [inline util/util.go:438]:
```suggestion
func GetCRDWithContract(ctx context.Context, c client.Reader, gvk schema.GroupVersionKind, contract string) (*apiextensionsv1.CustomResourceDefinition, error) {
```

Let's do a breaking change here? The new interface is a bit more strict which is a valuable thing to have, also going from `client.Client` to `client.Reader` is technically a compatible change, given that every Client is also a Reader

Comment (sbueringer) [reply util/util.go]: Sounds good - done. 

WDYT is there a better way to handle the breaking behavioral change in the parent func too? (https://github.com/kubernetes-sigs/cluster-api/pull/5698#discussion_r752054269)

File: util/conversion/conversion.go (hunk 1 above)
Comment (vincepri) [reply util/conversion/conversion.go:101]: I'm fine doing a breaking change here, given the implications of caching entire CRDs

Comment (sbueringer) [reply util/conversion/conversion.go:101]: Okay then let's keep it simple regarding deprecation and rotating through func names. I'll do a breaking change now to add the apiReader and we'll do another one once we don't support the fallback for the wrong CRD names anymore.

Comment (sbueringer) [reply util/conversion/conversion.go]: Done

Comment (vincepri) [reply util/conversion/conversion.go]: /retitle ⚠️  Update references with an APIReader as fallback

**Comment by sbueringer** [comment]
/retitle ⚠️ Update references with an APIReader as fallback

**Review by vincepri — APPROVED** [review]
/approve
/lgtm

**Pull request merged**